app/controllers/api/v1/users_controller.rb

```diff
@@ -29,6 +29,48 @@ def update
         end
       end
 
+      # For social media logins, renders the appropriate `redirect_to` path depending on whether the user is registered.
+      #
+      # Requires that the ActionController::Parameters contain a :user key, with a nested :email key.
+      # For example: { user: { email: "john@example.com" } }
+      #
+      # @return [String] A string of the user's redirect_to path
+      # @see https://github.com/zquestz/omniauth-google-oauth2#devise
+      #
+      def exist
+        user = User.find_by(email: params[:user][:email])
+
+        @redirect_path = '/profile'
+        unless user
+          @redirect_path = '/social_login'
+        end
+        render json: { redirect_to: @redirect_path }
+      end
+
+      # For social media logins, creates the user in the database if necessary,
+      # then logs them in, and renders the appropriate `redirect_to` path depending on whether the user is logging in
+      # for the first time.
+      #
+      # @return [String] A string of the user's redirect_to path
+      # @return [Json] A serialied JSON object derived from current_user
+      # @return [Token] A token that the frontend stores to know the user is logged in
+      # @see https://github.com/OperationCode/operationcode_backend/blob/master/app/controllers/api/v1/sessions_controller.rb#L8-L20
+      #
+      def social
+          set_social_user_and_redirect
+
+          if @user.save
+            sign_in @user, event: :authenticate_user
+            render json: {
+              token: @user.token,
+              user: UserSerializer.new(current_user),
+              redirect_to: @redirect_path
+            }
+          else
+            redirect_to new_user_registration_url, alert: @user.errors.full_messages.join("\n")
+          end
+      end
+
```

Comment:
A best practice I often hear DHH, and many others, recommending that **all** controller actions be RESTful.  Meaning, trying to use only CRUD actions (`index/create/new/edit/show/update/destroy`), unless you cannot do so.  

This would require an architectural change in that these two new controller actions would go into a new, dedicated controller.  Regarding the naming of the controller, you'd try to come up with one where the standard RESTful actions  make sense.

Perhaps something like `social_users_controller.rb`.

In this way, the `exist` action could be the `show` action.  And the `social` action could be the `create` action.

And then you could change the one-off routes to be resourceful routes (i.e. `resources :social_users, only: [:create, :show]`